File: Algorithms_and_Data_structures/ADS_2021/w4_BST/lab4/leetcode_additional/5_96.py
class Solution:
    def numTrees(self, n: int) -> int:
        
        arr = [1] * (n + 1)
        for nodes in range(2, n + 1):
            summ = 0
            for root in range(1, nodes + 1):
                left = root - 1 
                right = nodes - root 
                summ = summ + (arr[left] * arr[right])
            arr[nodes] = summ 
        
        return arr[n]

File: Algorithms_and_Data_structures/ADS_2021/w4_BST/lab4/leetcode_additional/5_96.cpp
class Solution {
public:
    int numTrees(int n) {
        int a[10000];
        for (int i = 0; i < n; i++){
            a[i] = 1;
        }

        for (int nodes = 2; nodes <= n; nodes++){
            int sum = 0;
            for (int root = 1; root <= nodes; root++){
                int left = root - 1;
                int right = nodes - root; 

                sum = sum + (a[left] * a[right]);
            }
            a[nodes] = sum;
        }

        return a[n];
        
    }
};